### ISSUE TITLE
Roslyn async-streams tests fail on Core

### ISSUE THREAD

**Issue by jcouv** [dotnet/coreclr#20032]
I'm working on async-streams for C# 8.0. Part of the feature is async-iterator methods, which allow both `yield` statements and `await` expressions to be used. Under the cover, we generate a state machine like for regular async methods, but also make use of `ManualResetValueTaskSourceLogic` and `IValueTaskSource`.

The problem is that all our tests for this feature work on the desktop runtime, but hang on Core runtime.  It looks like all the threads are waiting, with no progress made.
It's worth noting that the generated IL was verified to work outside of the test harness (xunit, AssemblyLoadContext).
@jaredpar, @jeffschwMSFT and I tried debugging to get a clue what is going on, but we were not able to find out what the threads are waiting for.

We need some help to investigate this issue. Thanks
Below are the details for repro'ing.

Repro steps:
1.	Unzip the repro folder from `\\jcouvdev\shared-dumps\async-iterators-coreclr\repro.zip`
2.	Open a command prompt in that folder
3.	Run `repro.cmd` (note this requires a connection to corpnet, as it will access some binaries from my nuget cache, which I was not able to bundle into the repro folder)

Outcome:
```
repro> repro.cmd

xUnit.net Console Runner (64-bit .NET Core 4.6.26614.01)
  Discovering: Microsoft.CodeAnalysis.CSharp.Emit.UnitTests
  Discovered:  Microsoft.CodeAnalysis.CSharp.Emit.UnitTests
  Starting:    Microsoft.CodeAnalysis.CSharp.Emit.UnitTests
… xunit hangs …
```

Expected:
Test should complete and succeed.

Some background:
For this repro, I picked one test (AsyncIteratorWithAwaitCompletedAndOneYieldAndOneInvocation), which compiles and executes the code below.
The repro uses the `2.0.7` runtime.

```C#
using static System.Console;
class C
{
    static async System.Collections.Generic.IAsyncEnumerable<int> M() // this code gets compiled into C.<M>d__0.MoveNext()
    {
        Write(""1 "");
        await System.Threading.Tasks.Task.CompletedTask;
        Write(""2 "");
        yield return 3;
        Write(""4 "");
    }
    static async System.Threading.Tasks.Task Main()
    {
        Write(""0 "");
        foreach await (var i in M())
        {
            Write($""{i} "");
        }
        Write(""Done"");
    }
}
```
The result of this compilation is available in the `binaries` sub-folder, as the binary with the GUID name.

**Comment by jkotas** [comment]
The test hang here: https://github.com/xunit/xunit/blob/9d10262a3694bb099ddd783d735aba2a7aac759d/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs#L268

XUnit's [`AsyncTestSyncContext::operationCount`](https://github.com/xunit/xunit/blob/9d10262a3694bb099ddd783d735aba2a7aac759d/src/xunit.execution/Sdk/AsyncTestSyncContext.cs#L16
) is 1. It means there was OperationStarted called without matching OperationCompleted. These methods are called the guts of AsyncMethodBuildBuilder.

Your code is basically doing the following - this will reproduce the same hang:

```csharp
[Fact]
public void Test1()
{
     AsyncVoidMethodBuilder.Create();
}
```

Where is the async operation in your test supposed to be completed? Step through that and find out why it is not calling `AsyncMethodBuildBuilder.OperationCompleted`.

**Comment by jkotas** [comment]
If it helps, here is where the async operation is started:

```
xunit_execution_dotnet!Xunit.Sdk.AsyncTestSyncContext.OperationStarted()
System_Private_CoreLib!System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Create()+0x30
c61b672a_c2ef_43f8_a2b7_13099c6ce0a0!C.M()+0x68
c61b672a_c2ef_43f8_a2b7_13099c6ce0a0!C+<Main>d__1.MoveNext()+0x96
System_Private_CoreLib!System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[[System.__Canon, System.Private.CoreLib]](System.__Canon ByRef)+0x49
c61b672a_c2ef_43f8_a2b7_13099c6ce0a0!C.Main()+0xa0
c61b672a_c2ef_43f8_a2b7_13099c6ce0a0!C.<Main>()+0x31
System_Private_CoreLib!System.Reflection.MethodBase.Invoke(System.Object, System.Object[])+0x22
Roslyn_Test_Utilities!Roslyn.Test.Utilities.CoreClr.TestExecutionLoadContext+<>c__DisplayClass7_0.<Execute>b__0()+0x1b1
```

**Comment by jcouv** [comment]
That is super helpful. Jared and I had traced to the xunit's `TestInvoker`, but hadn't managed to figure what it was waiting for.
I indeed do not generate any call to `SetResult` on the `AsyncVoidMethodBuilder` that is created in the state machine. That's because nobody is waiting for that result. I didn't realize that would have a negative impact.
I'll go ahead and close the issue. I'll revisit the generate code tomorrow, with this additional information.
Thanks!
